File: DSA_Problems/ArrayHashing/ContainsDuplicate/contains_duplicate.py

# https://leetcode.com/problems/contains-duplicate/


"""
1. Create set

2. Iterate over list/string, check if value already present in set
    If so: return true

3. If no duplicates found, return false outside loop


"""


def contains_duplicate_sort(nums):
    # Time: O(nlogn)
    # Space: O(1)
    nums.sort()

    for i in range(len(nums)-1):
        if nums[i] == nums[i+1]:
            return True
    return False


def contains_duplicate(nums):
    # Time O(N)
    # Space O(N)

    unique = set()

    for num in nums:
        if num in unique:
            return True
        else:
            unique.add(num)
    return False


if __name__ == '__main__':
    test1 = [1, 2, 3, 1]
    test2 = [1, 2, 3, 4]
    print(contains_duplicate(test1))
    print(contains_duplicate(test2))
    print(contains_duplicate_sort(test1))
    print(contains_duplicate_sort(test2))

File: DSA_Problems/ArrayHashing/ContainsDuplicate/containsDuplicate.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

bool containsDuplicate(vector<int> &nums)
{
    unordered_set<int> s;

    for (int i = 0; i < nums.size(); i++)
    {
        if (s.find(nums[i]) != s.end())
        {
            return true;
        }
        s.insert(nums[i]);
    }

    return false;
}

int main()
{
    vector<int> test1{1, 2, 3, 1};
    vector<int> test2{1, 2, 3, 4};
    cout << containsDuplicate(test1) << endl;
    cout << containsDuplicate(test2) << endl;
}